### PR TITLE
docs: Add migrate stytch guide

### PR DESCRIPTION
**Pull request by sixhobbits**
Depends on

- https://github.com/FusionAuth/fusionauth-contrib/pull/14
- https://github.com/FusionAuth/fusionauth-import-scripts/pull/13

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1206274755704511

### CODE REVIEW

File: astro/src/content/docs/lifecycle/migrate-users/bulk/stytch.mdx [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+
+import InlineUIElement from 'src/components/InlineUIElement.astro';
+import InlineField from 'src/components/InlineField.astro';
+import Aside from '/src/components/Aside.astro';
+import {RemoteCode} from '@fusionauth/astro-components';
+import AdditionalSupport from 'src/content/docs/lifecycle/migrate-users/bulk/_additional-support.mdx';
+
+## Overview
+
+This article explains how to import users from Stytch into FusionAuth. It is a low-level, technical tutorial focusing on transferring password hashes, calling APIs, and preparing data. The article does not explain the high-level process of a migration strategy. This is explained in the [migration overview article](/docs/lifecycle/migrate-users/bulk/general). Before proceeding, please ensure you understand concepts like dual write, backfill, final synchronization, rolling migration, and cutover. The [Stytch documentation on migration strategies](https://stytch.com/docs/guides/migrations/migrating-user-data-statically) is also useful reading. Be aware of all laws regarding the protection and transfer of personal information in your country.
```
Comment (mark-robustelli) [inline astro/src/content/docs/lifecycle/migrate-users/bulk/stytch.mdx:20]: Should these be part of the Prerequisites?  Where can I go to learn about these concepts if I am not familiar with them? No that I follow the Stytch doc link I see they are there, but should be make more clear. Also I think info is in regards to importing users to Stytch. I don't think we use those terms in migrating to FusionAuth. Do we need them here?

File: astro/src/content/docs/lifecycle/migrate-users/bulk/stytch.mdx [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+
+Finally, if you want to import user passwords in addition to user personal details, you need a basic understanding of how password hashing and salts work. FusionAuth has a [hashing article](/articles/security/math-of-password-hashing-algorithms-entropy) that is a good starting point.
+
+### Limitations
+
+- This article explains how to migrate Stytch users from a Consumer Authentication project. It has not been tested with their B2B Saas Authentication project type.
+- This article explains how to import passwords into FusionAuth, but does not deal with other Stytch authentication types: magic links, passkeys, passcodes, mobile biometrics, two-factor authentication, and social logins such as Google OAuth.
```
Comment (mark-robustelli) [inline astro/src/content/docs/lifecycle/migrate-users/bulk/stytch.mdx:27]: We do have a component that talks about social logins
import SocialLoginNote from 'src/content/docs/lifecycle/migrate-users/bulk/_social-login-note.mdx';

There are a few other components you may want to include. Please check out the Cognito migration guide and take a look at some of the following:
MappingUserAttributes
OtherEntitiesIntro
Identifiers
and others.

File: astro/src/content/docs/lifecycle/migrate-users/bulk/stytch.mdx [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+You cannot download your users' password hashes from Stytch using their API. To get them, email support@stytch.com from the email address you used to sign up with Stytch. Ask support to send your users' hashes to you.
+
+### Email The Request
+
+Stytch will encrypt the hashes with your public key. Attach this key in a `.pem` file to the email you send Stytch.
+
+To create a private and public key pair, use the commands below in a terminal.
+
+```shell
+openssl genpkey -algorithm RSA -out private_key.pem &&
```
Comment (mark-robustelli) [inline astro/src/content/docs/lifecycle/migrate-users/bulk/stytch.mdx:81]: Formatting issue:

Screenshot 2024-02-22 at 9 21 44 PM

File: astro/src/content/docs/lifecycle/migrate-users/bulk/stytch.mdx [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+If you don't want to send Stytch an email, you can use the example response files for the rest of this tutorial. The two files Stytch emailed are in `exampleData/2_emailResponse`. Just remember to change the sample user Ids to your actual user Ids as you proceed.
+</Aside>
+
+### Decrypt The Reply
+
+Stytch will reply with two files: An encrypted password hash file (`stytch-project-test-36510638-652a-4d3d-9a94-f0a7106582fc-hashes-2021-01-11.enc`) and the key to decrypt it (`key.bin.enc`).
+
+Decrypt the password file with the commands below.
+
+```shell
```
Comment (mark-robustelli) [inline astro/src/content/docs/lifecycle/migrate-users/bulk/stytch.mdx:99]: most of these code snippets have the same leading space on the first line issue.

File: astro/src/content/docs/lifecycle/migrate-users/bulk/stytch.mdx [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+
+## Install The Scrypt Password Hash Plugin For FusionAuth
+
+The scrypt hashing algorithm is not [natively supported by FusionAuth](/docs/reference/password-hashes). However, FusionAuth allows [custom plugins](/docs/extend/code/password-hashes/custom-password-hashing). There is a scrypt plugin accompanying this article [in this GitHub repository](https://github.com/FusionAuth/fusionauth-contrib/tree/master/Password%20Hashing%20Plugins).
+
+<Aside type="note">
+The linked repository has multiple plugin projects. You need to build and install only the hashing project.
+</Aside>
+
+- Download and unzip the repository, or use Git in a terminal with the code below.
```
Comment (mark-robustelli) [inline astro/src/content/docs/lifecycle/migrate-users/bulk/stytch.mdx:181]: This section is indented. Is that intentional?

File: astro/src/content/docs/lifecycle/migrate-users/bulk/stytch.mdx [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+  ```bash
+  git clone https://github.com/fusionauth/fusionauth-contrib.git
+  ```
+
+Open the file `fusionauth-contrib/Password Hashing Plugins/src/main/java/com/mycompany/fusionauth/plugins/ExampleStytchScryptPasswordEncryptor.java`. The content of the file is shown below.
+
+<RemoteCode
```
Comment (mark-robustelli) [inline astro/src/content/docs/lifecycle/migrate-users/bulk/stytch.mdx:188]: showing 404 code found error. Is this something that needs to be merged?

**Review by mark-robustelli — CHANGES_REQUESTED** [review]
@sixhobbits, I took a first pass and mentioned a few things. Overall it appears to be on the right path but the formatting is a little off.  Please take a look at our [other migration guides](http://localhost:3000/docs/lifecycle/migrate-users/bulk) and trying to get this in a similar format. Specifically: Overview, Prerequisites, Planning Considerations,Exporting Users, Importing User, Bulk Migration, What to Do Next and Additional support.

There are a few other you can use for various sections. Please check out the Cognito migration guide and take a look at some of the following:
MappingUserAttributes
OtherEntitiesIntro
Identifiers
and others.

When you are done reformatting, I will then run through all the code.

File: astro/src/content/docs/lifecycle/migrate-users/bulk/stytch.mdx (hunk 6 above)
Comment (mark-robustelli) [reply astro/src/content/docs/lifecycle/migrate-users/bulk/stytch.mdx:188]: So is that PR ready for review as well?

Comment (sixhobbits) [reply astro/src/content/docs/lifecycle/migrate-users/bulk/stytch.mdx:188]: @mark-robustelli - yes that one is also ready for review. We usually just mention any other scripts/repos etc as a dependency in the top-level comment in the main PR to fusionauth-site as we did here, but let me know if there's a better way to do this

File: astro/src/content/docs/lifecycle/migrate-users/bulk/stytch.mdx [hunk 7]
```diff
@@ -0,0 +1,381 @@
...
+
+Check the fourth column to ensure that only scrypt users are included. The first four columns are the most important: `id`, `hash`, `salt`, and `hash_method`.
+
+The `3_getUserDetails.mjs` script loops through each user in the `hash.csv` file, uses the API to get the user details from Stytch, and saves them to `js/users.json`.
+
+- Open `3_getUserDetails.mjs` and set your project Id and secret.
+- Run the code below in a terminal in the `fusionauth-import-scripts/stytch/js` directory.
+  ```sh
+  npm update
+  node 3_getUserDetails.mjs
```
Comment (mark-robustelli) [inline astro/src/content/docs/lifecycle/migrate-users/bulk/stytch.mdx:199]: In 1_makeStytchUsers.mjs project_id and secret:

project_id: 
secret: 

and here you use

projectId: 
secretKey: 

Is there a reason or can you update and standardize in scripts and language used here.

File: astro/src/content/docs/lifecycle/migrate-users/bulk/stytch.mdx [hunk 8]
```diff
@@ -0,0 +1,381 @@
...
+Open your hash file from Stytch in a text editor and remove all the header lines so only one user per row remains. Save this file to `stytch/js/hash.csv`. The file should look like `exampleData/4_hashFilePreparation/hash.csv` now.
+
+<RemoteCode
+    url={frontmatter.importCodeRoot + "/stytch/exampleData/4_hashFilePreparation/hash.csv"}
+    lang="csv"
+/>
+
+Check the fourth column to ensure that only scrypt users are included. The first four columns are the most important: `id`, `hash`, `salt`, and `hash_method`.
+
+The `3_getUserDetails.mjs` script loops through each user in the `hash.csv` file, uses the API to get the user details from Stytch, and saves them to `js/users.json`.
```
Comment (mark-robustelli) [inline astro/src/content/docs/lifecycle/migrate-users/bulk/stytch.mdx:193]: Need to be more clear they need to update the location of the hash.csv file if it is not in the same directory.

File: astro/src/content/docs/lifecycle/migrate-users/bulk/stytch.mdx [hunk 9]
```diff
@@ -0,0 +1,381 @@
...
+
+<RemoteCode
+    url={frontmatter.javaCodeRoot + "/Password Hashing Plugins/src/main/java/com/mycompany/fusionauth/plugins/ExampleStytchScryptPasswordEncryptor.java"}
+    lang="java"
+/>
+
+This program takes a password and salt and returns a hash. It accepts and returns the `+` form of Base64, but changes it to `-` to work with Java internally. The `factor` parameter is ignored. Scrypt instead has multiple parameters set at the top of the file. If any of these differ from the ones you received in your encrypted CSV file header, you need to change them in the Java file.
+
+There are two other files linked to this scrypt plugin that you shouldn't need to alter:
+- `fusionauth-contrib/Password Hashing Plugins/src/main/java/com/mycompany/fusionauth/plugins/ExampleStytchScryptPasswordEncryptorTest.java`, which holds unit tests for the hasher.
```
Comment (mark-robustelli) [inline astro/src/content/docs/lifecycle/migrate-users/bulk/stytch.mdx:242]: Should this be 'fusionauth-contrib/Password Hashing Plugins/src/test' ?

File: astro/src/content/docs/lifecycle/migrate-users/bulk/stytch.mdx [hunk 10]
```diff
@@ -0,0 +1,381 @@
...
+
+### Set Up FusionAuth And Deploy The Plugin
+
+Now copy `fusionauth-example-password-encryptor-0.1.0.jar` to your FusionAuth `plugins` directory and restart FusionAuth.
+
+If you are not already running FusionAuth or want to test this process on another instance, you can start FusionAuth in Docker.
+- Open a new terminal in the `fusionauth-import-scripts` directory and run the code below.
+  ```sh
+  cd stytch/fusionAuthDockerFiles
+  docker compose up
```
Comment (mark-robustelli) [inline astro/src/content/docs/lifecycle/migrate-users/bulk/stytch.mdx:273]: Stuck here.  Get Error Message: 

fa     | 2024-03-07 10:36:42.339 PM INFO  com.inversoft.maintenance.search.ElasticsearchSilentModeWorkflowTask - Attempt 100 of 100. Waiting for search to become available [SERVER_DOWN]. Error [http://search:9200].

File: astro/src/content/docs/lifecycle/migrate-users/bulk/stytch.mdx [hunk 11]
```diff
@@ -0,0 +1,381 @@
...
+  ```
+- FusionAuth will now be running and browsable at `http://localhost:9011`. You can log in with `admin@example.com` and `password`. The container is called `fa`.
+- In the `fusionauth-contrib` terminal where you built the plugin, run the commands below to copy the JAR file into the FusionAuth container `plugins` directory.
+  ```sh
+  docker exec fa mkdir /usr/local/fusionauth/plugins
```
Comment (mark-robustelli) [inline astro/src/content/docs/lifecycle/migrate-users/bulk/stytch.mdx:278]: I used started another instance of FusionAuth and tried this in the terminal window where I built the plugin.  Git get an error: bash: docker: command not found

Screenshot 2024-03-07 at 3 43 02 PM

**Review by mark-robustelli — CHANGES_REQUESTED** [review]
I was doing pretty good until trying to install the Scrypt Hash Plugin. 

Neither the docker method or the copy method are working for me.  Please provide guidance and I can continue review.

Also I see several Vale errors.  Is that something I need to correct on our end?

File: astro/src/content/docs/lifecycle/migrate-users/bulk/stytch.mdx [hunk 12]
```diff
@@ -0,0 +1,382 @@
...
+
+### Save The User Details And Hash To FusionAuth
+
+Now you have the users file `users.json` and the scrypt plugin is installed. To import the users into FusionAuth, you need to run the Node.js import script.
+
+Open a terminal in the `fusionauth-import-scripts` directory and run the code below.
+```sh
+cd stytch/js
+npm update
+node 4_import.mjs
```
Comment (mark-robustelli) [inline astro/src/content/docs/lifecycle/migrate-users/bulk/stytch.mdx:292]: I am getting the following errors when running this script with the example [users.json](https://github.com/FusionAuth/fusionauth-import-scripts/blob/master/stytch/exampleData/5_userDetailAndHashPreparation/users.json) file. The api key seems good.

Screenshot 2024-03-14 at 5 56 55 PM

**Review by mark-robustelli — CHANGES_REQUESTED** [review]
Moving forward.  I am now getting an error when importing the example data. If you want to set up some time to go through this, just let me know.  It is a high priority for me now.

File: astro/src/content/docs/lifecycle/migrate-users/bulk/stytch.mdx (hunk 12 above)
Comment (mark-robustelli) [reply astro/src/content/docs/lifecycle/migrate-users/bulk/stytch.mdx:292]: PEBCAK here.  Rebooted and running.

File: astro/src/content/docs/lifecycle/migrate-users/bulk/stytch.mdx [hunk 13]
```diff
@@ -0,0 +1,382 @@
...
+<Aside type="tip">
+You might want to set the user's registration and tenant Id in your production migration.
+```js
+faUser.registrations = Array<UserRegistration>;
+faUser.tenantId = UUID;
+```
+</Aside>
+
+
+### Use the Script
```
Comment (mark-robustelli) [inline astro/src/content/docs/lifecycle/migrate-users/bulk/stytch.mdx:306]:
```suggestion
```

File: astro/src/content/docs/lifecycle/migrate-users/bulk/stytch.mdx [hunk 14]
```diff
@@ -0,0 +1,382 @@
...
+  docker exec fa mkdir /usr/local/fusionauth/plugins
+  docker cp "Password Hashing Plugins/target/fusionauth-example-password-encryptor-0.1.0.jar" fa:/usr/local/fusionauth/plugins/fusionauth-example-password-encryptor-0.1.0.jar
+  ```
+- Finally, restart FusionAuth for it to detect the plugin. In the terminal where FusionAuth is running in Docker, press Ctrl + C to stop it, wait, and run `docker compose up` again.
+
+### Save The User Details And Hash To FusionAuth
+
+Now you have the users file `users.json` and the scrypt plugin is installed. To import the users into FusionAuth, you need to run the Node.js import script.
+
+Open a terminal in the `fusionauth-import-scripts` directory and run the code below.
```
Comment (mark-robustelli) [inline astro/src/content/docs/lifecycle/migrate-users/bulk/stytch.mdx:288]:
```suggestion
### Use the Script

Open a terminal in the `fusionauth-import-scripts` directory and run the code below.
```

File: astro/src/content/docs/lifecycle/migrate-users/bulk/stytch.mdx (hunk 12 above)
Comment (mark-robustelli) [reply astro/src/content/docs/lifecycle/migrate-users/bulk/stytch.mdx:292]: I am using the right version, I was just waiting to approve until I could run through and verify that it was working.

**Review by mark-robustelli — CHANGES_REQUESTED** [review]
Made it all the way through.  Merged the other PR now that I was able to run though it. A few more small changes and then is should be ready to go.

**Review by mark-robustelli — APPROVED** [review]
LGTM

**Pull request merged**